src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageSource.java

```diff
@@ -0,0 +1,660 @@
+		return this.groupId;
+	}
+
+	/**
+	 * Set the group.id property for the consumer.
+	 * @param groupId the group id.
+	 */
+	public void setGroupId(String groupId) {
+		this.groupId = groupId;
+	}
+
+	protected String getClientId() {
+		return this.clientId;
+	}
+
+	/**
+	 * Set the client.id property for the consumer.
+	 * @param clientId the client id.
+	 */
+	public void setClientId(String clientId) {
+		this.clientId = clientId;
+	}
+
+	protected long getPollTimeout() {
+		return this.pollTimeout;
+	}
+
+	/**
+	 * Set the pollTimeout for the poll() operations; default 50ms.
+	 * @param pollTimeout the poll timeout.
+	 */
+	public void setPollTimeout(long pollTimeout) {
+		this.pollTimeout = pollTimeout;
+	}
+
+	protected RecordMessageConverter getMessageConverter() {
+		return this.messageConverter;
+	}
+
+	/**
+	 * Set the message converter to replace the default
+	 * {@link MessagingMessageConverter}.
+	 * @param messageConverter the converter.
+	 */
+	public void setMessageConverter(RecordMessageConverter messageConverter) {
+		this.messageConverter = messageConverter;
+	}
+
+	protected Type getPayloadType() {
+		return this.payloadType;
+	}
+
+	/**
+	 * Set the payload type.
+	 * Only applies if a type-aware message converter is provided.
+	 * @param payloadType the type to convert to.
+	 */
+	public void setPayloadType(Type payloadType) {
+		this.payloadType = payloadType;
+	}
+
+	protected ProducerFactory<K, V> getTransactionalProducerFactory() {
+		return this.producerFactory;
+	}
+
+	/**
+	 * Set a transactional producer factory.
+	 * @param producerFactory the producer factory.
+	 */
+	public void setTransactionalProducerFactory(ProducerFactory<K, V> producerFactory) {
+		Assert.isTrue(producerFactory.transactionCapable(), "The producer factory must be transaction capable");
+		this.producerFactory = producerFactory;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "kafka:message-source";
+	}
+
+	@Override
+	protected synchronized Object doReceive() {
+		if (this.consumer == null) {
+			createConsumer();
+		}
+		ConsumerRecord<K, V> record;
+		TopicPartition topicPartition;
+		synchronized (this.consumerMonitor) {
+			Set<TopicPartition> paused = this.consumer.paused();
+			if (paused.size() > 0) {
+				this.consumer.resume(paused);
+			}
+			ConsumerRecords<K, V> records = this.consumer.poll(this.pollTimeout);
+			this.consumer.pause(this.partitions);
+			if (records == null || records.count() == 0) {
+				return null;
+			}
+			record = records.iterator().next();
+			topicPartition = new TopicPartition(record.topic(), record.partition());
+			if (records.count() > 1) {
+				this.consumer.seek(topicPartition, record.offset() + 1);
+			}
+		}
+		Producer<K, V> producer = null;
+		AtomicBoolean myTransaction = new AtomicBoolean();
+		AtomicReference<Producer<K, V>> producerRef = new AtomicReference<>();
+		KafkaAckInfo<K, V> ackInfo = new KafkaAckInfo<K, V>(this.consumerMonitor, this.groupId, this.consumer, record,
```

Comment:
Wouldn't the logic be easier if we stop having this class as `static`?

Reply:
Unfortunately, we can't make `KafkaAckCallback` non-static, so this won't work.

Reply:
Nor sure what is the problem to leave `KafkaAckCallback` as `static` for possible extension and use non-static dependant inner class... I might agree when I will understand better the reason of this overcomplication with manual transactions support you've introduced....

Reply:
You can't reference a non-static (KafkaAckInfo) from a static inner class (KafkaAckCallback) - I resolved it by adding an interface.

Thanks for pushing me on this one.

Reply:
Ah! I see now. Well, some kind of Java peculiarity.
I have moved the `KafkaAckCallback` to the top level and now I can reference to the non-static inner class. I wonder what is the difference in this case between `static` inner class and top level one... In the end we definitely still can create their instances...

Sorry for the noise though.
I was just confused with those atomic references variables and properties when we would be able to get access to the `KafkaAckInfo` properties directly without static modifier. At least that is my feeling.